File: sem4/ai/lista3/player/game_client.go
/*
Based on game_client.c by Maciej Gębala


*/

package main

import (
	"fmt"
	"net"
	"os"
	"player/eval"
	"strconv"
	"strings"
)

func main() {
	var evaluationMode int = 1
	if len(os.Args) == 7 {
		evaluationMode, _ = strconv.Atoi(os.Args[6])
		fmt.Println("Debug mode enabled: evaluation mode", evaluationMode)
	} else if len(os.Args) != 6 {
		fmt.Println("Wrong number of arguments")
		os.Exit(1)
	}

	eval.SetEvaluationMode(evaluationMode)

	serverIP := os.Args[1]
	serverPort := os.Args[2]
	playerStr := os.Args[3]
	gameID := os.Args[4]
	depth := os.Args[5]
	// Parse player number
	player, err := strconv.Atoi(playerStr)
	if err != nil {
		fmt.Printf("Invalid player number: %v\n", err)
		os.Exit(1)
	}

	depthInt, err := strconv.Atoi(depth)
	if err != nil {
		fmt.Printf("Invalid depth: %v\n", err)
		os.Exit(1)
	}

	// Connect to server
	conn, err := net.Dial("tcp", net.JoinHostPort(serverIP, serverPort))
	if err != nil {
		fmt.Printf("Unable to connect: %v\n", err)
		os.Exit(1)
	}
	defer conn.Close()
	fmt.Println("Connected with server successfully")

	// Receive initial server message
	buffer := make([]byte, 16)
	n, err := conn.Read(buffer)
	if err != nil {
		fmt.Printf("Error while receiving server's message: %v\n", err)
		os.Exit(1)
	}
	serverMessage := strings.TrimSpace(string(buffer[:n]))
	fmt.Printf("Received: %s\n", serverMessage)

	// Send player information
	playerMessage := fmt.Sprintf("%s %s", playerStr, gameID)
	_, err = conn.Write([]byte(playerMessage))
	if err != nil {
		fmt.Printf("Unable to send message: %v\n", err)
		os.Exit(1)
	}

	var board eval.Board
	// Initialize board
	board.SetEmptyBoard()
	endGame := false

	// Main game loop
	for !endGame {
		// Receive server message
		buffer = make([]byte, 16)
		n, err := conn.Read(buffer)
		if err != nil {
			fmt.Printf("Error while receiving server's message: %v\n", err)
			os.Exit(1)
		}

		serverMessage = strings.TrimSpace(string(buffer[:n]))
		msg, err := strconv.Atoi(serverMessage)
		if err != nil {
			fmt.Printf("Error parsing server message: %v\n", err)
			continue
		}

		// Extract move and message code
		move := msg % 100
		msgCode := msg / 100

		// Update board with opponent's move
		if move != 0 {
			board.SetMove(move, 3-player)
		}

		// Handle message codes
		if msgCode == 0 || msgCode == 6 {
			myMove := eval.Move(board, depthInt, player)
			board.SetMove(myMove, player)

			// Send move to server
			moveStr := strconv.Itoa(myMove)
			_, err = conn.Write([]byte(moveStr))
			if err != nil {
				fmt.Printf("Unable to send message: %v\n", err)
				os.Exit(1)
			}
		} else {
			// Game ended
			endGame = true
			switch msgCode {
			case 1:
				fmt.Println("You won.")
			case 2:
				fmt.Println("You lost.")
			case 3:
				fmt.Println("Draw.")
			case 4:
				fmt.Println("You won. Opponent error.")
			case 5:
				fmt.Println("You lost. Your error.")
			}
		}
	}
}

File: sem4/ai/lista3/example/game_client.c
/****************************
Maciej Gębala (CC BY-NC 4.0)
Client ver. 0.2
2025-04-13
****************************/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>

#include "./board.h"


int main(int argc, char *argv[]) {
  int server_socket;
  struct sockaddr_in server_addr;
  char server_message[16], player_message[16];

  bool end_game;
  int player, msg, move;

  if ( argc != 5 ) {
    printf("Wrong number of arguments\n");
    return -1;
  }

  // Create socket
  server_socket = socket(AF_INET, SOCK_STREAM, 0);
  if ( server_socket < 0 ) {
    printf("Unable to create socket\n");
    return -1;
  }
  printf("Socket created successfully\n");

  // Set port and IP the same as server-side
  server_addr.sin_family = AF_INET;
  server_addr.sin_port = htons(atoi(argv[2]));
  server_addr.sin_addr.s_addr = inet_addr(argv[1]);

  // Send connection request to server
  if ( connect(server_socket, (struct sockaddr*)&server_addr, sizeof(server_addr)) < 0 ) {
    printf("Unable to connect\n");
    return -1;
  }
  printf("Connected with server successfully\n");

  // Receive the server message
  memset(server_message, '\0', sizeof(server_message));
  if ( recv(server_socket, server_message, sizeof(server_message), 0) < 0 ) {
    printf("Error while receiving server's message\n");
    return -1;
  }
  printf("Server message: %s\n", server_message);

  memset(player_message, '\0', sizeof(player_message));
  snprintf(player_message, sizeof(player_message), "%s %s", argv[3], argv[4]);
  // Send the message to server
  if ( send(server_socket, player_message, strlen(player_message), 0) < 0 ) {
    printf("Unable to send message\n");
    return -1;
  }

  setBoard();
  end_game = false;
  sscanf(argv[3], "%d", &player);

  while ( !end_game ) {
    memset(server_message, '\0', sizeof(server_message));
    if ( recv(server_socket, server_message, sizeof(server_message), 0) < 0 ) {
      printf("Error while receiving server's message\n");
      return -1;
    }
    printf("Server message: %s\n", server_message);
    sscanf(server_message, "%d", &msg);
    move = msg%100;
    msg = msg/100;
    if ( move != 0 ) {
      setMove(move, 3-player);
      printBoard();
    }
    if ( (msg == 0) || (msg == 6) ) {
      printf("Your move: ");
      if ( fgets(player_message, sizeof(player_message), stdin) == NULL ) {
        printf("Error while reading move\n");
        return -1;
      }
      sscanf(player_message, "%d", &move);
      setMove(move, player);
      printBoard();
      memset(player_message, '\0', sizeof(player_message));
      snprintf(player_message, sizeof(player_message), "%d", move);
      if ( send(server_socket, player_message, strlen(player_message), 0) < 0 ) {
        printf("Unable to send message\n");
        return -1;
      }
      printf("Player message: %s\n", player_message);
     } else {
       end_game = true;
       switch ( msg ) {
         case 1 : printf("You won.\n"); break;
         case 2 : printf("You lost.\n"); break;
         case 3 : printf("Draw.\n"); break;
         case 4 : printf("You won. Opponent error.\n"); break;
         case 5 : printf("You lost. Your error.\n"); break;
      }
    }
  }

  // Close socket
  close(server_socket);

  return 0;
}
